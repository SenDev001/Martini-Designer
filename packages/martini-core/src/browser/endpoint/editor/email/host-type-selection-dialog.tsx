import * as React from "react";
import { BaseDialog, applySize } from "../../../dialogs/dialogs";
import messages from "martini-messages/lib/messages";
import { ListItem, List } from "../../../components/list";
import { HostType } from "./host-type";

export class HostTypeSelectionDialog extends BaseDialog<HostType>{
    value: HostType;

    constructor(private readonly hostTypes: HostType[]) {
        super({
            title: messages.select_host_type_title
        });
        this.value = this.hostTypes[0];
        applySize(this.contentNode, {
            height: 280
        });

        this.appendCloseButton(messages.cancel_btn);
        this.appendAcceptButton(messages.select_btn);
    }

    protected doRender(): React.ReactNode {
        const items: ListItem[] = this.hostTypes && this.hostTypes.map((type, i) => ({
            label: `${type.name}`,
            data: type,
            selected: i === 0
        }));

        return this.hostTypes && <List
            items={items}
            filtered={true}
            style={{ height: "100%" }}
            focus={true}
            onSelectionChanged={selection => {
                this.value = selection.data;
                this.validate();
            }}
            onDoubleClick={() => this.accept()}
        />;
    }
}
