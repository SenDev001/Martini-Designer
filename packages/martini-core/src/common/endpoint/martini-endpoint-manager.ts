import { JsonRpcServer } from "@theia/core";
import { MartiniEvent } from "../event/martini-event";

export const martiniEndpointManagerPath = "/services/martini/endpoints";

export enum EndpointType {
    DIR_WATCHER = "dir-watcher",
    EMAIL = "email",
    FTP_CLIENT = "ftp-client",
    FTP_SERVER = "ftp-server",
    HTTP_FILTER = "http-filter",
    JABBER = "jabber",
    JMS_LISTENER = "jms-listener",
    KAFKA_LISTENER = "kafka-listener",
    MONGODB_LISTENER = "mongodb-listener",
    URL_ALIAS = "rest-alias",
    RSS = "rss",
    SCHEDULER = "scheduler",
    TRACKER_RESUBMIT = "tracker-resubmit",
    REDIS_LISTENER = "redis-listener"
}

export interface MartiniEndpoint {
    type: EndpointType;
    name: string;
    service: string;
    packageName: string;
    enabled: boolean;
    modifiable: boolean;
    documentType: string;
    track: boolean;
    replicated: boolean;
    status: "STARTED" | "STOPPED";
}

export namespace MartiniEndpoint {
    export function is(object: any): object is MartiniEndpoint {
        return (
            !!object &&
            typeof object === "object" &&
            "type" in object &&
            "name" in object &&
            "service" in object &&
            "packageName" in object &&
            "enabled" in object &&
            "modifiable" in object &&
            "status" in object
        );
    }
    export function getType(type: string): EndpointType | undefined {
        const typeKey: string | undefined = Object.keys(EndpointType).find(
            (key: string) => EndpointType[key as keyof typeof EndpointType] === type
        );
        return typeKey ? EndpointType[typeKey as keyof typeof EndpointType] : undefined;
    }
}

export const MartiniEndpointManager = Symbol("MartiniEndpointManager");

export interface MartiniEndpointManager extends JsonRpcServer<MartiniEndpointManagerClient> {
    getAll(packageName: string): Promise<MartiniEndpoint[]>;

    get(packageName: string, endpointName: string): Promise<MartiniEndpoint | undefined>;

    start(packageName: string, endpointName: string): Promise<void>;

    stop(packageName: string, endpointName: string): Promise<void>;

    delete(packageName: string, endpointName: string): Promise<void>;

    save(packageName: string, endpoint: MartiniEndpoint): Promise<void>;

    test(packageName: string, endpoint: MartiniEndpoint): Promise<void>;
}

export interface EndpointEvent extends MartiniEvent {
    name: string;
    packageName: string;
    event: "STARTED" | "STOPPED" | "DELETED" | "SAVED";
}

export namespace EndpointEvent {
    export function is(event: MartiniEvent): event is EndpointEvent {
        return event.type === "ENDPOINT";
    }
}

export interface MartiniEndpointManagerClient {
    onEvent(event: EndpointEvent): void;
}

export const getDisplayName = (type: EndpointType): string => {
    switch (type) {
        case EndpointType.DIR_WATCHER:
            return "Directory Watcher";
        case EndpointType.EMAIL:
            return "Email";
        case EndpointType.KAFKA_LISTENER:
            return "Kafka Listener";
        case EndpointType.SCHEDULER:
            return "Scheduler";
        case EndpointType.TRACKER_RESUBMIT:
            return "Tracker Resubmit";
        case EndpointType.JABBER:
            return "Jabber";
        case EndpointType.FTP_CLIENT:
            return "FTP Client";
        case EndpointType.FTP_SERVER:
            return "FTP Server";
        case EndpointType.JMS_LISTENER:
            return "JMS Listener";
        case EndpointType.RSS:
            return "RSS";
        case EndpointType.URL_ALIAS:
            return "URL Alias";
        case EndpointType.MONGODB_LISTENER:
            return "MongoDB Listener";
        case EndpointType.REDIS_LISTENER:
            return "Redis Pub/Sub Listener";
        case EndpointType.HTTP_FILTER:
            return "HTTP Filter";
        default:
            return "Unknown";
    }
};

export interface RssEndpoint extends MartiniEndpoint {
    schedule: string;
    onlyNew: boolean;
    rssUrl: string;
}

export interface EmailSettings {
    host: string;
    port: number;
    username: string;
    password: string;
    ssl: boolean;
}

export interface ReplyEmailSettings extends EmailSettings {
    from: string;
}

export interface EmailEndpoint extends MartiniEndpoint, EmailSettings {
    schedule: string;
    replicated: boolean;
    track : boolean;
    protocol: "imap" | "pop3";
    deleteOnReceive: boolean;
    sendOutputAsReply: boolean;
    sendReplyOnError: boolean;
    replyEmailSettings: ReplyEmailSettings;
}
