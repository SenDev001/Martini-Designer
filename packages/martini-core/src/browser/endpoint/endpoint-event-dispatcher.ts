import {Emitter, Event} from "@theia/core";
import {injectable} from "inversify";
import {EndpointEvent, MartiniEndpointManagerClient} from "../../common/endpoint/martini-endpoint-manager";

export const EndpointEventDispatcher = Symbol("EndpointEventDispatcher");

export interface EndpointEventDispatcher {
    onEndpointEvent: Event<EndpointEvent>;
}

@injectable()
export class EndpointEventDispatcherClient
    implements MartiniEndpointManagerClient, EndpointEventDispatcher {
    private readonly onEndpointEventEmitter = new Emitter<EndpointEvent>();
    readonly onEndpointEvent = this.onEndpointEventEmitter.event;

    onEvent(event: EndpointEvent): void {
        this.onEndpointEventEmitter.fire(event);
    }
}
