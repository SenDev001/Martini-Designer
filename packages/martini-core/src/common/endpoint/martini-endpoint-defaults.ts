import { EndpointType, MartiniEndpoint, RssEndpoint , EmailEndpoint } from "./martini-endpoint-manager";

export const createDefaultEndpoint = (packageName: string, name: string, type: EndpointType): MartiniEndpoint => {
    switch (type) {
        case EndpointType.RSS:
            return createDefaultRssEndpoint(packageName, name);
        case EndpointType.EMAIL:
            return createDefaultEmailEndpoint(packageName, name);
        default:
            return {
                name,
                type,
                packageName,
                status: "STOPPED",
                enabled: true
            } as MartiniEndpoint;
    }
};

const createDefaultRssEndpoint = (packageName: string, name: string): RssEndpoint => ({
    name,
    type: EndpointType.RSS,
    packageName,
    service: "",
    modifiable: true,
    replicated: true,
    status: "STOPPED",
    documentType: "RSS",
    enabled: true,
    track: false,
    schedule: "repeating:1000",
    rssUrl: "",
    onlyNew: false
});

const createDefaultEmailEndpoint = (packageName: string, name: string): EmailEndpoint => ({
    name,
    type: EndpointType.EMAIL,
    packageName,
    service: "",
    modifiable: true,
    replicated: true,
    status: "STOPPED",
    documentType: "E-Mail",
    enabled: true,
    track: false,
    schedule: "repeating:1000",
    host: "",
    port: 993,
    username : "",
    password : "",
    deleteOnReceive : false ,
    sendReplyOnError : false ,
    sendOutputAsReply : false ,
    protocol : "imap" ,
    ssl : false,
    replyEmailSettings : {
        host: "",
        port: 1,
        username: "",
        password: "",
        ssl: false,
        from: ""
    }
});
