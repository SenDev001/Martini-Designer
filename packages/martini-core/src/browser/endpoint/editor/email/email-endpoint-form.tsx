import { Field, Form, Formik, FormikErrors, useFormikContext } from "formik";
import { noop } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import Collapsible from "react-collapsible";
import * as Yup from "yup";
import { EmailEndpoint, MartiniEndpoint, ReplyEmailSettings } from "../../../../common/endpoint/martini-endpoint-manager";
import { DocumentType } from "../../../../common/tracker/document-type-manager";
import { FormRow, OnFormChange, validateSchema } from "../../../form/form";
import "../../../yup/yup-ext";
import { GeneralConfigurationSection, generalSchema } from "../general-config-section";
import { HostType, hostTypes, replyHostTypes } from "./host-type";
import { HostTypeSelectionDialog } from './host-type-selection-dialog';

export interface EmailEndpointFormProps {
    endpoint: MartiniEndpoint;
    reset: boolean;
    documentTypeProvider: () => Promise<DocumentType[]>;
    onChange: (connection: MartiniEndpoint) => void;
    onValidate: (errors: FormikErrors<MartiniEndpoint>) => void;
}

export const EmailEndpointForm: React.FC<EmailEndpointFormProps> = ({
    endpoint,
    reset,
    documentTypeProvider,
    onChange,
    onValidate
}) => {
    const schema = React.useMemo(
        () =>
            Yup.object()
                .shape<Partial<EmailEndpoint>>({
                    host: Yup.string().required(),
                    port: Yup.number().min(1).max(65535).required(),
                    username: Yup.string()
                        .required()
                        .email(),
                    schedule: Yup.string()
                        .required()
                        .min(1)
                        .test("is-positive-integer", messages.must_be_greater_or_equal(1), value => {
                            if (value) {
                                const index = value.lastIndexOf(":");
                                if (index > -1) {
                                    const sub = value.substring(index + 1, value.length + 1);
                                    const interval = Number.parseInt(sub);
                                    if (!isNaN(interval) && interval > 0) return true;
                                }
                            }

                            return false;
                        }),
                    replyEmailSettings: Yup.object<ReplyEmailSettings>().when(["sendOutputAsReply", "sendReplyOnError"], {
                        is: (sendOutputAsReply, sendReplyOnError) => sendOutputAsReply || sendReplyOnError,
                        then: Yup.object().shape<Partial<ReplyEmailSettings>>({
                            host: Yup.string().required(),
                            port: Yup.number().min(1).max(65535).required(),
                            username: Yup.string().email(),
                            from: Yup.string().email()
                        }),
                    })
                })
                .concat(generalSchema),
        []
    );

    const validate = async (endpoint: EmailEndpoint) => validateSchema(schema, endpoint, onValidate);

    return (
        <Formik
            initialValues={endpoint}
            initialTouched={{
                host: true,
                username: true,
                port: true,
                schedule: true,
                replyEmailSettings: {
                    host: true,
                    port: true,
                    username: true,
                    from: true
                }
            }}
            enableReinitialize={reset}
            validate={validate}
            validateOnMount={true}
            onSubmit={noop}
        >
            <Form>
                <OnFormChange onChange={onChange} />
                <GeneralConfigurationSection documentTypeProvider={documentTypeProvider} />
                <EmailConfigurationSection />
                <EmailActionsSection />
            </Form>
        </Formik>
    );
};

const EmailConfigurationSection: React.FC = () => {
    const [opened, setOpened] = React.useState(true);
    const handleFocus = () => setOpened(true);
    const context = useFormikContext<EmailEndpoint>();

    const handleScheduleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const interval = Number.parseInt(e.currentTarget.value);
        if (isNaN(interval)) context.setFieldValue("schedule", `repeating:${e.currentTarget.value}`);
        else context.setFieldValue("schedule", `repeating:${interval * 1000}`);
    };

    const getInterval = (schedule: string) => {
        const index = schedule.lastIndexOf(":");
        if (index > -1) {
            const sub = schedule.substring(index + 1, schedule.length + 1);
            const interval = Number.parseInt(sub);
            if (!isNaN(interval)) return (interval / 1000).toString();
            return sub;
        }
        return "0";
    };

    const hostTypeInputRef = React.useRef<HTMLInputElement>();
    const handleBrowseHostTypes = async () => {
        const dlg = new HostTypeSelectionDialog(hostTypes);
        const hostType = await dlg.open();
        if (hostType) {
            context.setFieldValue("host", hostType.host);
            context.setFieldValue("port", hostType.port);
            context.setFieldValue("username",
                HostType.isValidUsername(context.values.username, hostType) ?
                    context.values.username :
                    HostType.convertUsername(context.values.username, hostType)
            );
            context.setFieldValue("ssl", hostType.ssl);
            context.setFieldValue("protocol", hostType.protocol);
            hostTypeInputRef.current?.focus();
        }
    };

    return (
        <Collapsible
            trigger={messages.configuration_title("Email")}
            open={opened}
            onClosing={() => setOpened(false)}
            onOpening={() => setOpened(true)}
        >
            <div
                style={{
                    display: "grid",
                    gridColumnGap: 10,
                    gridTemplateColumns: "max-content 1fr"
                }}
                onFocus={handleFocus}
            >
                <FormRow
                    name="host"
                    label={messages.host}
                    tooltip={messages.email_host_tooltip}
                    gridTemplateColumns="1fr max-content"
                >
                    <Field name="host" type="text" className="theia-input" />
                    <input
                        type="button"
                        className="theia-button"
                        value={messages.choose}
                        onClick={handleBrowseHostTypes}
                    />
                </FormRow>
                <FormRow
                    name="port"
                    label={messages.host}
                    tooltip={messages.email_port_tooltip}
                    defaultValue={993}>
                    <Field
                        name="port"
                        type="number"
                        className="theia-input"
                        min="1"
                        max="65535"
                    />
                </FormRow>
                <FormRow
                    name="username"
                    label={messages.username}
                    tooltip={messages.email_username_tooltip}
                >
                    <Field name="username" type="text" className="theia-input" />
                </FormRow>
                <FormRow
                    name="password"
                    label={messages.password}
                    tooltip={messages.email_password_tooltip}
                >
                    <Field name="password" type="password" className="theia-input" />
                </FormRow>
                <FormRow
                    name="schedule"
                    label={messages.email_polling_interval_seconds}
                    tooltip={messages.email_polling_interval_tooltip}
                    defaultValue={1}
                >
                    <Field
                        name="schedule"
                        type="number"
                        className="theia-input"
                        min="1"
                        value={getInterval(context.values.schedule)}
                        onChange={handleScheduleChange}
                    />
                </FormRow>
                <FormRow
                    name="ssl"
                    label="SSL"
                    tooltip={messages.email_ssl_tooltip}
                    defaultValue={false}>
                    <Field name="ssl" type="checkbox" />
                </FormRow>
                <FormRow
                    name="protocol"
                    label={messages.protocol}
                    tooltip={messages.email_protocol_tooltip}
                    defaultValue="imap"
                >
                    <label>
                        <Field name="protocol" type="radio" value="imap" className="theia-input" /> IMAP
                    </label>

                    <label>
                        <Field name="protocol" type="radio" value="pop3" className="theia-input" /> POP3
                    </label>
                </FormRow>
            </div>
        </Collapsible>
    );
};

const EmailActionsSection: React.FC = () => {
    const [opened, setOpened] = React.useState(true);
    const handleFocus = () => setOpened(true);
    const context = useFormikContext<EmailEndpoint>();
    return (
        <Collapsible
            trigger={messages.email_actions}
            open={opened}
            onClosing={() => setOpened(false)}
            onOpening={() => setOpened(true)}
        >
            <div
                style={{
                    display: "grid",
                    gridColumnGap: 10,
                    gridTemplateColumns: "max-content 1fr"
                }}
                onFocus={handleFocus}
            >
                <FormRow
                    name="deleteOnReceive"
                    label={messages.delete_email_once_received}
                    tooltip={messages.email_deleteOnReceive_tooltip}
                >
                    <Field name="deleteOnReceive" type="checkbox" />
                </FormRow>
                <FormRow
                    name="sendOutputAsReply"
                    label={messages.send_service_response_as_reply}
                    tooltip={messages.email_sendOutputAsReply_tooltip}
                >
                    <Field name="sendOutputAsReply" type="checkbox" />
                </FormRow>
                <FormRow
                    name="sendReplyOnError"
                    label={messages.send_reply_on_error}
                    tooltip={messages.email_sendReplyOnError_tooltip}
                >
                    <Field name="sendReplyOnError" type="checkbox" />
                </FormRow>
            </div>
            {(context.values.sendOutputAsReply || context.values.sendReplyOnError) && <EmailReplySettings />}
        </Collapsible>
    );
};

const EmailReplySettings: React.FC = () => {
    const context = useFormikContext<EmailEndpoint>();
    const hostTypeInputRef = React.useRef<HTMLInputElement>();
    const handleBrowseHostType = async () => {
        const dlg = new HostTypeSelectionDialog(replyHostTypes);
        const hostType = await dlg.open();
        if (hostType) {
            context.setFieldValue("replyEmailSettings.host", hostType.host);
            context.setFieldValue("replyEmailSettings.port", hostType.port);
            context.setFieldValue("replyEmailSettings.username",
                HostType.isValidUsername(context.values.replyEmailSettings.username, hostType) ?
                    context.values.replyEmailSettings.username :
                    HostType.convertUsername(context.values.replyEmailSettings.username, hostType)
            );
            context.setFieldValue("replyEmailSettings.ssl", hostType.ssl);
            context.setFieldValue("replyEmailSettings.from", hostType.from);
            hostTypeInputRef.current?.focus();
        }
    };
    return (
        <div
            style={{
                display: "grid",
                gridColumnGap: 10,
                gridTemplateColumns: "max-content 1fr",
                margin: "0px 12px"
            }}
        >
            <FormRow
                name="replyEmailSettings.host"
                label={messages.host}
                tooltip={messages.email_port_tooltip}
                gridTemplateColumns="1fr max-content"
            >
                <Field name="replyEmailSettings.host" type="text" className="theia-input" />
                <input
                    type="button"
                    className="theia-button"
                    value={messages.choose}
                    onClick={handleBrowseHostType}
                />
            </FormRow>
            <FormRow
                name="replyEmailSettings.port"
                label={messages.port}
                tooltip={messages.email_port_tooltip}
                defaultValue={993}
            >
                <Field name="replyEmailSettings.port" type="number" className="theia-input" />
            </FormRow>
            <FormRow
                name="replyEmailSettings.username"
                label={messages.username}
                tooltip={messages.email_username_tooltip}
            >
                <Field name="replyEmailSettings.username" type="text" className="theia-input" />
            </FormRow>
            <FormRow
                name="replyEmailSettings.password"
                label={messages.password}
                tooltip={messages.email_password_tooltip}
            >
                <Field name="replyEmailSettings.password" type="password" className="theia-input" />
            </FormRow>
            <FormRow
                name="replyEmailSettings.from"
                label={messages.from}
                tooltip={messages.email_from_tooltip}
            >
                <Field name="replyEmailSettings.from" type="text" className="theia-input" />
            </FormRow>
            <FormRow name="replyEmailSettings.ssl" label="SSL/TLS" tooltip={messages.email_ssl_tooltip}>
                <Field name="replyEmailSettings.ssl" type="checkbox" />
            </FormRow>
        </div>
    );
};
