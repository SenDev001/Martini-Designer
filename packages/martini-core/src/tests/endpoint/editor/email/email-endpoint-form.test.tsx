require("reflect-metadata");
import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import { noop } from "lodash";
import * as React from "react";
import * as Renderer from "react-test-renderer";
import { EmailEndpointForm } from "../../../../browser/endpoint/editor/email/email-endpoint-form";
import { createDefaultEndpoint } from "../../../../common/endpoint/martini-endpoint-defaults";
import { EndpointType, EmailEndpoint } from "../../../../common/endpoint/martini-endpoint-manager";

Enzyme.configure({ adapter: new Adapter() });

const endpoint = createDefaultEndpoint("test", "test", EndpointType.EMAIL) as EmailEndpoint;

test("EmailEndpointForm should be rendered", () => {
    const component = Renderer.create(<EmailEndpointForm
        endpoint={endpoint}
        documentTypeProvider={async () => []}
        onChange={noop}
        onValidate={noop}
        reset={false}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
