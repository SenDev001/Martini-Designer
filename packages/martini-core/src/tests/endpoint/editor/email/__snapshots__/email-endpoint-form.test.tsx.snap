// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailEndpointForm should be rendered 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="Collapsible"
  >
    <span
      className="Collapsible__trigger is-open"
      onClick={[Function]}
      onKeyPress={[Function]}
      style={null}
      tabIndex={null}
    >
      General Configuration
    </span>
    <div
      className="Collapsible__contentOuter"
      onTransitionEnd={[Function]}
      style={
        Object {
          "WebkitTransition": "none",
          "height": "auto",
          "msTransition": "none",
          "overflow": "hidden",
          "transition": "none",
        }
      }
    >
      <div
        className="Collapsible__contentInner"
      >
        <div
          onFocus={[Function]}
          style={
            Object {
              "display": "grid",
              "gridColumnGap": 10,
              "gridTemplateColumns": "max-content 1fr",
            }
          }
        >
          <label
            htmlFor="service"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The service to execute when the endpoint is triggered."
          >
            Service
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr max-content",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              className="theia-input"
              name="service"
              onBlur={[Function]}
              onChange={[Function]}
              readOnly={true}
              type="text"
              value=""
            />
            <input
              className="theia-button"
              type="button"
              value="Browse..."
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="documentType"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The document type to be registered to Tracker when this endpoint is triggered."
          >
            Document Type
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr max-content",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              className="theia-input"
              name="documentType"
              onBlur={[Function]}
              onChange={[Function]}
              readOnly={true}
              type="text"
              value="E-Mail"
            />
            <input
              className="theia-button"
              onClick={[Function]}
              type="button"
              value="Browse..."
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="enabled"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="Whether or not the endpoint should be started automatically when the package is started."
          >
            Auto Start
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              checked={true}
              name="enabled"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={true}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="track"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="Flag determining whether executions should be logged to Tracker."
          >
            Log to Tracker
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              checked={false}
              name="track"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={false}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="replicated"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="If this endpoint is configured on a Martini instance that's running in a cluster, replicated will determine whether to run the endpoint on all instances, or only the elected leader node in the cluster. When this is checked, all instances will run the endpoint. When it's unchecked, only the leader node will run the endpoint."
          >
            Replicated
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              checked={true}
              name="replicated"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={true}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="Collapsible"
  >
    <span
      className="Collapsible__trigger is-open"
      onClick={[Function]}
      onKeyPress={[Function]}
      style={null}
      tabIndex={null}
    >
      Email Configuration
    </span>
    <div
      className="Collapsible__contentOuter"
      onTransitionEnd={[Function]}
      style={
        Object {
          "WebkitTransition": "none",
          "height": "auto",
          "msTransition": "none",
          "overflow": "hidden",
          "transition": "none",
        }
      }
    >
      <div
        className="Collapsible__contentInner"
      >
        <div
          onFocus={[Function]}
          style={
            Object {
              "display": "grid",
              "gridColumnGap": 10,
              "gridTemplateColumns": "max-content 1fr",
            }
          }
        >
          <label
            htmlFor="host"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The email server to connect to."
          >
            Host
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr max-content",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              className="theia-input"
              name="host"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <input
              className="theia-button"
              onClick={[Function]}
              type="button"
              value="Choose"
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="port"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The server port number."
          >
            Host
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              className="theia-input"
              max="65535"
              min="1"
              name="port"
              onBlur={[Function]}
              onChange={[Function]}
              type="number"
              value={993}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="username"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The email address used to login."
          >
            Username
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              className="theia-input"
              name="username"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="password"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The password used to login."
          >
            Password
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              className="theia-input"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              type="password"
              value=""
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="schedule"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The interval in seconds, at which Martini checks the inbox."
          >
            Polling Interval (seconds)
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              className="theia-input"
              min="1"
              name="schedule"
              onBlur={[Function]}
              onChange={[Function]}
              type="number"
              value="1"
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="ssl"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="Flag to determine whether the connection uses SSL."
          >
            SSL
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              checked={false}
              name="ssl"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={false}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="protocol"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The email protocol to use. Supports protocols: IMAP and POP3."
          >
            Protocol
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <label>
              <input
                checked={true}
                className="theia-input"
                name="protocol"
                onBlur={[Function]}
                onChange={[Function]}
                type="radio"
                value="imap"
              />
               IMAP
            </label>
            <label>
              <input
                checked={false}
                className="theia-input"
                name="protocol"
                onBlur={[Function]}
                onChange={[Function]}
                type="radio"
                value="pop3"
              />
               POP3
            </label>
            <div
              className="form-error-message-wrapper"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="Collapsible"
  >
    <span
      className="Collapsible__trigger is-open"
      onClick={[Function]}
      onKeyPress={[Function]}
      style={null}
      tabIndex={null}
    >
      Email Actions
    </span>
    <div
      className="Collapsible__contentOuter"
      onTransitionEnd={[Function]}
      style={
        Object {
          "WebkitTransition": "none",
          "height": "auto",
          "msTransition": "none",
          "overflow": "hidden",
          "transition": "none",
        }
      }
    >
      <div
        className="Collapsible__contentInner"
      >
        <div
          onFocus={[Function]}
          style={
            Object {
              "display": "grid",
              "gridColumnGap": 10,
              "gridTemplateColumns": "max-content 1fr",
            }
          }
        >
          <label
            htmlFor="deleteOnReceive"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="Flag determining whether the received email be deleted after the service is invoked."
          >
            Delete Email Once Received
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              checked={false}
              name="deleteOnReceive"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={false}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="sendOutputAsReply"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="Flag determining whether the output of the service be used as a reply body."
          >
            Send Service Response As Reply
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              checked={false}
              name="sendOutputAsReply"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={false}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="sendReplyOnError"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="Flag determining whether the exception thrown by the service, if any, be sent back as reply body."
          >
            Send Reply On Error
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content max-content",
              }
            }
          >
            <input
              checked={false}
              name="sendReplyOnError"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={false}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
